man/date.man
DATE(1)                                            User Commands                                           DATE(1)

NAME
       date - print or set the system date and time

SYNOPSIS
       date [OPTION]... [+FORMAT]
       date [-u|--utc|--universal] [MMDDhhmm[[CC]YY][.ss]]

DESCRIPTION
       Display the current time in the given FORMAT, or set the system date.

       Mandatory arguments to long options are mandatory for short options too.

       -d, --date=STRING
              display time described by STRING, not 'now'

       --debug
              annotate the parsed date, and warn about questionable usage to stderr

       -f, --file=DATEFILE
              like --date; once for each line of DATEFILE

       -I[FMT], --iso-8601[=FMT]
              output  date/time  in  ISO 8601 format.  FMT='date' for date only (the default), 'hours', 'minutes',
              'seconds', or 'ns' for date and time to the indicated precision.  Example: 2006-08-14T02:34:56-06:00

       -R, --rfc-email
              output date and time in RFC 5322 format.  Example: Mon, 14 Aug 2006 02:34:56 -0600

       --rfc-3339=FMT
              output date/time in RFC 3339 format.  FMT='date', 'seconds', or 'ns' for date and time to the  indi‐
              cated precision.  Example: 2006-08-14 02:34:56-06:00

       -r, --reference=FILE
              display the last modification time of FILE

       -s, --set=STRING
              set time described by STRING

       -u, --utc, --universal
              print or set Coordinated Universal Time (UTC)

       --help display this help and exit

       --version
              output version information and exit

       FORMAT controls the output.  Interpreted sequences are:

       %%     a literal %

       %a     locale's abbreviated weekday name (e.g., Sun)

       %A     locale's full weekday name (e.g., Sunday)

       %b     locale's abbreviated month name (e.g., Jan)

       %B     locale's full month name (e.g., January)

       %c     locale's date and time (e.g., Thu Mar  3 23:05:25 2005)

       %C     century; like %Y, except omit last two digits (e.g., 20)

       %d     day of month (e.g., 01)

       %D     date; same as %m/%d/%y

       %e     day of month, space padded; same as %_d

       %F     full date; same as %Y-%m-%d

       %g     last two digits of year of ISO week number (see %G)

       %G     year of ISO week number (see %V); normally useful only with %V

       %h     same as %b

       %H     hour (00..23)

       %I     hour (01..12)

       %j     day of year (001..366)

       %k     hour, space padded ( 0..23); same as %_H

       %l     hour, space padded ( 1..12); same as %_I

       %m     month (01..12)

       %M     minute (00..59)

       %n     a newline

       %N     nanoseconds (000000000..999999999)

       %p     locale's equivalent of either AM or PM; blank if not known

       %P     like %p, but lower case

       %q     quarter of year (1..4)

       %r     locale's 12-hour clock time (e.g., 11:11:04 PM)

       %R     24-hour hour and minute; same as %H:%M

       %s     seconds since 1970-01-01 00:00:00 UTC

       %S     second (00..60)

       %t     a tab

       %T     time; same as %H:%M:%S

       %u     day of week (1..7); 1 is Monday

       %U     week number of year, with Sunday as first day of week (00..53)

       %V     ISO week number, with Monday as first day of week (01..53)

       %w     day of week (0..6); 0 is Sunday

       %W     week number of year, with Monday as first day of week (00..53)

       %x     locale's date representation (e.g., 12/31/99)

       %X     locale's time representation (e.g., 23:13:48)

       %y     last two digits of year (00..99)

       %Y     year

       %z     +hhmm numeric time zone (e.g., -0400)

       %:z    +hh:mm numeric time zone (e.g., -04:00)

       %::z   +hh:mm:ss numeric time zone (e.g., -04:00:00)

       %:::z  numeric time zone with : to necessary precision (e.g., -04, +05:30)

       %Z     alphabetic time zone abbreviation (e.g., EDT)

       By default, date pads numeric fields with zeroes.  The following optional flags may follow '%':

       -      (hyphen) do not pad the field

       _      (underscore) pad with spaces

       0      (zero) pad with zeros

       ^      use upper case if possible

       #      use opposite case if possible

       After  any  flags  comes  an optional field width, as a decimal number; then an optional modifier, which is
       either E to use the locale's alternate representations if available, or O to  use  the  locale's  alternate
       numeric symbols if available.

EXAMPLES
       Convert seconds since the epoch (1970-01-01 UTC) to a date

              $ date --date='@2147483647'

       Show the time on the west coast of the US (use tzselect(1) to find TZ)

              $ TZ='America/Los_Angeles' date

       Show the local time for 9AM next Friday on the west coast of the US

              $ date --date='TZ="America/Los_Angeles" 09:00 next Fri'

DATE STRING
       The  --date=STRING  is  a  mostly free format human readable date string such as "Sun, 29 Feb 2004 16:21:42
       -0800" or "2004-02-29 16:21:42" or even "next Thursday".  A date string may contain items indicating calen‐
       dar  date, time of day, time zone, day of week, relative time, relative date, and numbers.  An empty string
       indicates the beginning of the day.  The date string format is more complex than is easily documented  here
       but is fully described in the info documentation.

AUTHOR
       Written by David MacKenzie.

REPORTING BUGS
       GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
       Report date translation bugs to <http://translationproject.org/team/>

COPYRIGHT
       Copyright   ©   2017  Free  Software  Foundation,  Inc.   License  GPLv3+:  GNU  GPL  version  3  or  later
       <http://gnu.org/licenses/gpl.html>.
       This is free software: you are free to change and redistribute it.  There is NO  WARRANTY,  to  the  extent
       permitted by law.

SEE ALSO
       Full documentation at: <http://www.gnu.org/software/coreutils/date>
       or available locally via: info '(coreutils) date invocation'

GNU coreutils 8.28                                 January 2018                                            DATE(1)


php/Date.class.php
<?php
// ------------------------------------------------------------
// Date  (PHP Class)
//    |
//    |-> Simplifies Timestamp exporting/importing while keeping full microsecond precision AND timezone accuracy
//    |
//    |-> Replaces DateTime::RFC3339_EXTENDED (requires PHP 7+) with Date::RFC3339_MICROSECONDS (see below)
//    |
//    |-> Note: In PHP 7, there exists a bug while using RFC3339_EXTENDED in conjunction with DateTime::createFromFormat regarding the milliseconds ('v'), which microseconds ('u') avoids
//
if (class_exists('Date')===false) {
	class Date extends DateTime {

		const FRONTEND_DATE = "m/d/Y";
		//    08/15/2005                            "m/d/Y"

		const FRONTEND_DATE_LONG = "m/d/Y (l)";
		//    08/15/2005 (Monday)                   "m/d/Y (l)"

		const FRONTEND_DATETIME = "m/d/Y H:i:s";
		//    08/15/2005 15:52:01                   "m/d/Y H:i:s"

		const FRONTEND_DATETIME_LONG = "m/d/Y h:i:s A";
		//    08/15/2005 03:52:01 PM                "m/d/Y h:i:s A"

		const RFC3339_MICROSECONDS = "Y-m-d\TH:i:s.uP";
		//    2005-08-15T15:52:01.457896+00:00      "Y-m-d\TH:i:s.uP"

		const RFC2822_MICROSECONDS = "D, d M Y H:i:s.u O";
		//    Mon, 15 Aug 2005 15:52:01.457896 +0000   "D, d M Y H:i:s.u O"

		const FILENAME_TIMESTAMP = "Y-m-d.H-i-s";
		//    2005-08-15.15-52-01                   "Y-m-d.H-i-s"

		const FILENAME_TIMESTAMP_MICROSECONDS = "Y-m-d.H-i-s.u";
		//    2005-08-15.15-52-01.457896            "Y-m-d.H-i-s.u"
		
		const MYSQL_DATE = "Y-m-d";
		//    2005-08-15                            "Y-m-d"

		const MYSQL_DATETIME  = "Y-m-d H:i:s";
		const MYSQL_TIMESTAMP = "Y-m-d H:i:s";
		//    2005-08-15 15:52:01                   "Y-m-d H:i:s"

		// const RFC3339
		// const ATOM
		// const DATE_ATOM 
		// const W3C
		//    2005-08-15T15:52:01+00:00             "Y-m-d\TH:i:sP"
		
		// const RFC3339_EXTENDED
		//    2005-08-15T15:52:01.000+00:00         "Y-m-d\TH:i:s.vP"

		// const RFC1123
		// const RFC2822
		// const RSS
		//    Mon, 15 Aug 2005 15:52:01 +0000       "D, d M Y H:i:s O"

		// const RFC822
		// const RFC1036
		//    Mon, 15 Aug 05 15:52:01 +0000         "D, d M y H:i:s O"

		// const COOKIE
		//    Monday, 15-Aug-2005 15:52:01 UTC      "l, d-M-Y H:i:s T"

		// const RFC850
		//    Monday, 15-Aug-05 15:52:01 UTC        "l, d-M-y H:i:s T"
		
		// WARNING - DateTimeInterface::ISO8601 is not compatible with ISO-8601 - it was left this way for backward compatibility reasons. Use DateTime::RFC3339 or DateTime::ATOM, instad
		// const ISO8601
		//    2005-08-15T15:52:01+0000              "Y-m-d\TH:i:sO"

		// public function __construct(...) {
		// ...
		// 	PULLED FROM PARENT CLASS 'DateTime' (PHP native class)
		// ...
		// }

		// function getTimezoneOffset() {
		// 	...
		// 		PULLED FROM PARENT CLASS 'DateTime' (PHP native class)
		// 	...
		// }

		function getJsTimezoneOffset() {
			// TODO: Add code to mirror Javascript logic
		}

		function getFullYear() {
			// TODO: Add code to mirror Javascript logic
		}

		function getMonth() {
			// TODO: Add code to mirror Javascript logic
		}

		function getDate() {
			// TODO: Add code to mirror Javascript logic
		}

		function getHours() {
			// TODO: Add code to mirror Javascript logic
		}

		function getMinutes() {
			// TODO: Add code to mirror Javascript logic
		}

		function getSeconds() {
			// TODO: Add code to mirror Javascript logic
		}

	}
}


// ------------------------------------------------------------
// format_datetime  (PHP Function)
//   |
//   |--> Format a date coming from [ PHP-Class-Object ] format, and format is to [ MySQL Database (YYYY-mm-dd) ] format
//
if (!function_exists('format_datetime')) {
	function format_datetime($datetime_str="", $format_str=DATE_RFC3339, $set_timezone="UTC", $input_format="") {
		//	Numeric DateTime strings crash/error-out during PHP's default DateTime constructor 
		//	 |--> If we're given a numeric string without an associated input-format, then handle it as "EpochSeconds.DecimalSeconds" formatted-input
		if ((empty($input_format)===true) && (is_numeric($datetime_str)===true)) {
			if (is_decimal($datetime_str)===true) {
				// WITH decimal seconds, e.g. "1565893386.629922"
				$datetime_str = number_format(floatval($datetime_str), 6, '.', '');
				$input_format = "U.u";
			} else {
				// NO decimal seconds, e.g. "1565893386"
				$datetime_str = number_format(intval($datetime_str), 0, '', '');
				$input_format = "U";
			}
		}

		if ((empty($input_format)===false) && (is_string($input_format)===true)) {
			// Instantiate the DateTime object with a given date-string format ( see: https://www.php.net/manual/en/function.date.php )
			$timestamp_obj = Date::createFromFormat($input_format, $datetime_str);

		} elseif (in_array(strtolower($datetime_str),array("","now"))) {
			$timestamp_obj = new Date("now");

		} else {
			// is_decimal
			$timestamp_obj = new Date($datetime_str);

		}
		
		// Set the Input Timezone (if-empty)
		if (empty($set_timezone)===true) {
			$set_timezone = date_default_timezone_get();
		} elseif ((is_string($set_timezone)===true) && (in_array(strtolower($set_timezone),array("edt","est","server")))) {
			$set_timezone = date_default_timezone_get();
		}
		$timestamp_obj->setTimezone(new DateTimeZone($set_timezone));

		// Set the Output Formatting (if-empty)
		if (empty($format_str)===true) {
			$format_str=DATE_RFC3339;
		}

		// Return the final, formatted datetime string
		return ($timestamp_obj->format($format_str));
	}
}


// ------------------------------------------------------------
// format_date_to_mysql  (PHP Function)
//   |
//   |--> Formats a date coming from [ End-User Readable (mm/dd/YYYY) ] format, and format is to [ MySQL Database (YYYY-mm-dd) ] format
//
if (!function_exists('format_date_to_mysql')) {
	function format_date_to_mysql($orig_date, $format="Y-m-d") {
		// If formatted like 2007-02-04, then just return it.
		if (preg_match('/\A\d{4}-\d{2}-\d{2}\z/', $orig_date)) {
			$formatted_date = $orig_date;
		// If 3 Mar 2008 type, convert it -- Note: This format (dd mmm yyyy) is commonly used by the US Military
		} elseif (preg_match('%\b(0?[1-9]|[12][0-9]|3[01])[- /][A-Za-z]{3}[- /](19|20)?[0-9]{2}\b%m', $orig_date)) {
			$date_stamp = strtotime(str_replace(array("/","-"), " ", $orig_date));
			$formatted_date = date("Y-m-d", $date_stamp);
		} else {
			// Check for invalid date
			list($mm,$dd,$yy)=explode("/",str_replace("-", "/", $orig_date));
			if (not_null($orig_date) && checkdate(intval($mm),intval($dd),intval($yy))) {
				$formatted_date = str_pad($yy, 4, '200', STR_PAD_LEFT).'-'.str_pad($mm, 2, '0', STR_PAD_LEFT).'-'.str_pad($dd, 2, '0', STR_PAD_LEFT);
			} else {
				$formatted_date = "0000-00-00";
			}
		}
		return $formatted_date;
	}
}


// ------------------------------------------------------------
// format_date_from_mysql  (PHP Function)
//   |
//   |--> Format a date coming from [ MySQL Database (YYYY-mm-dd) ] format, and format is to [ End-User Readable (mm/dd/YYYY) ] format
//
if (!function_exists('format_date_from_mysql')) {
	function format_date_from_mysql($orig_date, $format="m/d/Y") {
		if (not_null($orig_date)) {
			$date_stamp = strtotime($orig_date);
			$formatted_date = date($format, $date_stamp);
		} else {
			$formatted_date = '';
		}
		return $formatted_date;
	}
}


// ------------------------------------------------------------
// Citation(s)
//
//   php.net  |  "date (format chars)"  |  https://www.php.net/manual/en/function.date.php
//
//   php.net  |  "DateTime (class)"  |  https://php.net/manual/en/class.datetime.php
//
//   php.net  |  "DateTime::createFromFormat"  |  https://php.net/manual/en/datetime.createfromformat.php
//
// ------------------------------------------------------------

?>